connector/connect/client/src/main/scala/org/apache/spark/sql/connect/client/SparkSession.scala

```diff
@@ -0,0 +1,133 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.client
+
+import scala.language.existentials
+
+import io.grpc.{ManagedChannel, ManagedChannelBuilder}
+import org.apache.arrow.memory.RootAllocator
+
+import org.apache.spark.SPARK_VERSION
+import org.apache.spark.connect.proto
+
+
+class SparkSession(
+    private val userContext: proto.UserContext,
+    private val channel: ManagedChannel)
+  extends AutoCloseable {
+  private[this] val stub = proto.SparkConnectServiceGrpc.newBlockingStub(channel)
+
+  private[this] val allocator = new RootAllocator()
+
+  /**
+   * The version of Spark on which this application is running.
+   */
+  def version: String = SPARK_VERSION
+
+  /**
+   * Returns a `DataFrame` with no rows or columns.
+   *
+   * @since 3.4.0
+   */
+  @transient
+  lazy val emptyDataFrame: Dataset = newDataset { builder =>
+    builder.getLocalRelationBuilder
+  }
+
+  /**
+   * Creates a [[Dataset]] with a single `LongType` column named `id`, containing elements
+   * in a range from `start` to `end` (exclusive) with a step value, with partition number
+   * specified.
+   *
+   * @since 2.0.0
+   */
+  def range(start: Long, end: Long, step: Long, numPartitions: Int): Dataset = {
+    range(start, end, step, Option(numPartitions))
+  }
+
+  private def range(start: Long, end: Long, step: Long, numPartitions: Option[Int]): Dataset = {
+    newDataset { builder =>
+      val rangeBuilder = builder.getRangeBuilder
+        .setStart(start)
+        .setEnd(end)
+        .setStep(step)
+      numPartitions.foreach(rangeBuilder.setNumPartitions)
+    }
+  }
+
+  /**
+   * Executes a SQL query using Spark, returning the result as a `DataFrame`.
+   * This API eagerly runs DDL/DML commands, but not for SELECT queries.
+   *
+   * @since 2.0.0
+   */
+  def sql(query: String): Dataset = newDataset { builder =>
+    builder.setSql(proto.SQL.newBuilder().setQuery(query))
+  }
+
+  private[client] def newDataset(f: proto.Relation.Builder => Unit): Dataset = {
+    val builder = proto.Relation.newBuilder()
+    f(builder)
+    val plan = proto.Plan.newBuilder().setRoot(builder).build()
+    new Dataset(this, plan)
+  }
+
+  private[client] def analyze(plan: proto.Plan): proto.AnalyzePlanResponse = {
+    val request = proto.AnalyzePlanRequest.newBuilder()
+      .setPlan(plan)
+      .setUserContext(userContext)
+      .build()
+    stub.analyzePlan(request)
+  }
+
+  override def close(): Unit = {
+    channel.shutdownNow()
+    allocator.close()
+  }
+}
+
+object SparkSession {
+  def builder(): Builder = new Builder()
+
+  class Builder() {
```

Comment:
It would be good to use a similar `ChannelBuilder` construct as in the Python client to have the same behavior on both sides. I understand this is kind of painful to do for the first PR. Maybe the alternative is to make the first PR even smaller.

Reply:
yeah support ChannelBuilder probably means that you should support https://github.com/apache/spark/blob/master/connector/connect/docs/client-connection-string.md

Reply:
We use a channel builder in the build method?

I also don't quite understand the comment about the connection string. That is just an option we should support (URI parsing is very much a solved problem).

Finally, sure we can make this PR smaller. It is early days for the Scala Client and we need to start somewhere.

Reply:
@hvanhovell that was the point with the suggestion to start the PR without the GRPC client at all.

Reply:
@amaliujia I mean it is option during configuration. I didn't mean it is optional. We will add it.

Reply:
IMO we are burning a lot of cycles on code that is just meant to check if the project properly compiles.